float64 x
float64 y
float64 z

float64 alpha
float64 beta
float64 gamma